File: java/src/main/java/com/company/automation/models/PagedProducts.java
package com.company.automation.models;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.List;

/**
 * The pagination envelope the catalogue returns.
 *
 * <p>{@code from} and {@code to} are nullable: on an empty result the API sends null for both,
 * which is exactly the case a pagination test needs to cover.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record PagedProducts(
    @JsonProperty("current_page") int currentPage,
    @JsonProperty("data") List<Product> data,
    @JsonProperty("from") Integer from,
    @JsonProperty("to") Integer to,
    @JsonProperty("last_page") int lastPage,
    @JsonProperty("per_page") int perPage,
    @JsonProperty("total") int total) {

  /** True when this page carries no items. */
  public boolean isEmpty() {
    return data == null || data.isEmpty();
  }

  /** Number of items on this page, 0 when the page is empty. */
  public int size() {
    if (data == null) {
      return 0;
    }
    return data.size();
  }

  /**
   * The ids on this page, in order. A plain loop rather than a stream chain: this is read by people
   * whose main language may not be Java.
   */
  public List<String> ids() {
    List<String> ids = new java.util.ArrayList<>();
    if (data == null) {
      return ids;
    }
    for (Product product : data) {
      ids.add(product.id());
    }
    return ids;
  }
}

File: csharp/src/AutomationFramework.Core/Models/PagedProducts.cs
using System.Text.Json.Serialization;

namespace AutomationFramework.Core.Models;

/// <summary>
/// The pagination envelope the catalogue returns.
/// </summary>
/// <remarks>
/// <c>From</c> and <c>To</c> are nullable: on an empty result the API sends null for both, which is
/// exactly the case a pagination test needs to cover.
/// </remarks>
public sealed record PagedProducts(
    [property: JsonPropertyName("current_page")] int CurrentPage,
    [property: JsonPropertyName("data")] IReadOnlyList<Product> Data,
    [property: JsonPropertyName("from")] int? From,
    [property: JsonPropertyName("to")] int? To,
    [property: JsonPropertyName("last_page")] int LastPage,
    [property: JsonPropertyName("per_page")] int PerPage,
    [property: JsonPropertyName("total")] int Total)
{
    /// <summary>True when this page carries no items.</summary>
    public bool IsEmpty => Data is null || Data.Count == 0;

    /// <summary>Number of items on this page, 0 when the page is empty.</summary>
    public int Size => Data?.Count ?? 0;

    /// <summary>
    /// The ids on this page, in order.
    /// </summary>
    /// <remarks>
    /// A plain loop rather than a LINQ projection: this is read by people whose main language may not be
    /// C#, and a loop is always understood.
    /// </remarks>
    public List<string> Ids()
    {
        List<string> ids = [];
        if (Data is null)
        {
            return ids;
        }

        foreach (Product product in Data)
        {
            ids.Add(product.Id);
        }

        return ids;
    }
}
